cql3/statements/list_effective_service_level_statement.cc

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "list_effective_service_level_statement.hh"
+#include "auth/role_manager.hh"
+#include "cql3/statements/prepared_statement.hh"
+#include <memory>
+#include "exceptions/exceptions.hh"
+#include "service/qos/qos_common.hh"
+#include "service/query_state.hh"
+#include "cql3/result_set.hh"
+#include "types/types.hh"
+#include "duration.hh"
+#include "transport/messages/result_message.hh"
+
+namespace cql3 {
+
+namespace statements {
+
+list_effective_service_level_statement::list_effective_service_level_statement(sstring role_name)
+: _role_name(std::move(role_name)) {}
+
+std::unique_ptr<prepared_statement> 
+list_effective_service_level_statement::prepare(data_dictionary::database db, cql_stats& stats) {
+    return std::make_unique<prepared_statement>(::make_shared<list_effective_service_level_statement>(*this));
+}
+
+static auto make_column(sstring name, const shared_ptr<const abstract_type> type) {
+    return make_lw_shared<column_specification>(
+        "QOS",
+        "effective_service_level",
+        ::make_shared<column_identifier>(std::move(name), true),
+        type);
+};
+
+static bytes_opt decompose_timeout (const qos::service_level_options::timeout_type& duration) {
+    return std::visit(overloaded_functor{
+        [&] (const qos::service_level_options::unset_marker&) {
+            return bytes_opt();
+        },
+        [&] (const qos::service_level_options::delete_marker&) {
+            return bytes_opt();
+        },
+        [&] (const lowres_clock::duration& d) -> bytes_opt {
+            auto nanos = std::chrono::duration_cast<std::chrono::nanoseconds>(d).count();
+            return utf8_type->decompose(to_string(cql_duration(months_counter{0}, days_counter{0}, nanoseconds_counter{nanos})));
+        },
+    }, duration);
+};
+
+future<::shared_ptr<cql_transport::messages::result_message>>
+list_effective_service_level_statement::execute(query_processor& qp, service::query_state& state, const query_options&) const {
+    static thread_local const std::vector<lw_shared_ptr<column_specification>> metadata({
+        make_column("service_level_option", utf8_type),
+        make_column("effective_service_level", utf8_type),
+        make_column("value", utf8_type)
+    });
+    auto& role_manager = state.get_client_state().get_auth_service()->underlying_role_manager();
+
+    if (!co_await role_manager.exists(_role_name)) {
+        throw auth::nonexistant_role(_role_name);
+    }
+
+    auto role_set = co_await role_manager.query_granted(_role_name, auth::recursive_role_query::yes);
```

Comment:
Did you verify that the `role_set` contains also the loged_in role itself (`_role_name`)?

Reply:
Yes
https://github.com/scylladb/scylladb/blob/master/auth/standard_role_manager.cc#L525